From 5a4ff0112a76a3c279ba9e6e753d192451aa55ba Mon Sep 17 00:00:00 2001
From: Brian Delwiche <delwiche@google.com>
Date: Tue, 1 Oct 2024 23:12:54 +0000
Subject: [PATCH] (Reland) Encrypt LE link immediately on reconnection

LE link must be encrypted immediately on connection if device are
already bonded.

This is a backport of ag/29056565, but the code needs to go in a
different location because that patch relies on recent feature work.

Ignore-AOSP-First: security
Test: mmm packages/modules/Bluetooth
Bug: 288144143
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:d254a228ebc58cb7eb1470faf506e5b722cea096)
Merged-In: Ib9075586ed08eceac917909f2c07d54f7702901f
Change-Id: Ib9075586ed08eceac917909f2c07d54f7702901f
---
 system/stack/acl/btm_acl.cc | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/system/stack/acl/btm_acl.cc b/system/stack/acl/btm_acl.cc
index b0145d5701..666beb9ee1 100644
--- a/system/stack/acl/btm_acl.cc
+++ b/system/stack/acl/btm_acl.cc
@@ -59,6 +59,7 @@
 #include "rust/src/core/ffi/types.h"
 #include "stack/acl/acl.h"
 #include "stack/acl/peer_packet_types.h"
+#include "stack/btm/btm_ble_sec.h"
 #include "stack/btm/btm_dev.h"
 #include "stack/btm/btm_int_types.h"
 #include "stack/btm/btm_sco.h"
@@ -1052,6 +1053,16 @@ void StackAclBtmAcl::btm_establish_continue(tACL_CONN* p_acl) {
           default_packet_type_mask, p_acl->RemoteAddress());
     }
     btm_set_link_policy(p_acl, btm_cb.acl_cb_.DefaultLinkPolicy());
+  } else if (p_acl->is_transport_ble()) {
+    tBTM_SEC_DEV_REC* p_dev_rec = btm_find_dev(p_acl->remote_addr);
+
+    if (p_dev_rec == nullptr) {
+      log::warn("No security record for {}", p_acl->RemoteAddress());
+    } else if (p_dev_rec->sec_rec.is_le_link_key_known()) {
+      btm_ble_set_encryption(
+          p_acl->remote_addr, BTM_BLE_SEC_ENCRYPT,
+          p_dev_rec->role_central ? HCI_ROLE_CENTRAL : HCI_ROLE_PERIPHERAL);
+    }
   }
   NotifyAclLinkUp(*p_acl);
 }
-- 
2.46.1.824.gd892dcdcdd-goog

